sdk/identity/Azure.Identity/src/UserAssertionScope.cs

```diff
@@ -0,0 +1,75 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Identity.Client;
+
+namespace Azure.Identity
+{
+    /// <summary>
+    /// Creates a scope which will be used by <see cref="OnBehalfOfCredential"/> to request tokens on behalf of the user's token used to initialize
+    /// the <see cref="UserAssertionScope"/> instance.
+    /// </summary>
+    public class UserAssertionScope : IDisposable
+    {
+        private static AsyncLocal<UserAssertionScope> _currentAsyncLocal = new();
+        internal UserAssertion UserAssertion { get; }
+        internal MsalConfidentialClient Client { get; set; }
+        internal ITokenCacheOptions CacheOptions { get; }
+
+        /// <summary>
+        /// The current value of the <see cref="AsyncLocal{UserAssertion}"/>.
+        /// </summary>
+        internal static UserAssertionScope Current => _currentAsyncLocal.Value;
+
+        /// <summary>
+        /// Initializes a new instance of <see cref="UserAssertionScope"/> using the supplied access token.
+        /// </summary>
+        /// <param name="accessToken">The access token that will be used by <see cref="OnBehalfOfCredential"/> as the user assertion when requesting On-Behalf-Of tokens.</param>
+        /// <param name="hydrateCache"> The delegate to be called which retrieves the cache from persistence for this user assertion partition. </param>
+        /// <param name="persistCache"> The delegate to be called with the current state of the token cache on each time it is updated for this <see cref="UserAssertionScope"/> instance. </param>
+        public UserAssertionScope(string accessToken, Func<Task<ReadOnlyMemory<byte>>> hydrateCache = null, Func<ReadOnlyMemory<byte>, Task> persistCache = null)
```

Comment:
Do we need both to be set at the same time? Can customer override just one?

Reply:
If they want hooks to hydrate and persist the cache, then yes. Neither are required, but if the intention is to manage the cache, both are required.

Reply:
That's a scary ctor. Should we express this via virtual members instead?

Reply:
Or an option type.

Reply:
@schaabs and I discussed making these delegates take options types also rather than being so specific with the inputs and return types. Wrapping the whole thing in another options type isn't a bad idea either.